Add plugins endpoints

TimothyBJacobs opened the pull request:
Introduces a set of plugin endpoints, intended to compliment the block directory searching endpoints.

Brings tests and endpoint code from tellyworth/wordpress-develop/add/rest-api/block-directory.

For some reason it seems like the filesystem isn't available when running tests.

## Description
<!-- Please describe what you have changed or added -->

## How has this been tested?
This has been manually tested and has some test coverage. Needs more test coverage and coverage on multisite.

## Types of changes
New Feature

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

TimothyBJacobs committed:
First pass at plugins endpoint.

Brings tests and endpoint code from tellyworth/wordpress-develop/add/rest-api/block-directory.

For some reason it seems like the filesystem isn't available when running tests.

TimothyBJacobs committed:
phpcs fixes. Don't check status permissions if not updating it

TimothyBJacobs committed:
Try forcing the filesystem method to direct

TimothyBJacobs committed:
Fix spacing issue

TimothyBJacobs commented:
Something I can't figure out is that it seems like I can't write to `WP_PLUGIN_DIR` during the tests. `WP_Filesystem()` also isn't available, `get_filesystem_method()` ends up returning `ftpext`. The tests pass locally for me by forcing `FS_METHOD` during the test, but this doesn't seem to work on travis.

Maybe related to #20090? Pinging @noahtallen.

noahtallen commented:
@TimothyBJacobs that's odd. We have access to https://github.com/wp-phpunit/wp-phpunit in the unit tests, and not all of wordpress-develop. Could that be it?

Travis and `npm run test-unit-php` should be working in the same environment (`wp-env`) 🤔

noahtallen commented:
Yeah, I had just chmoded the build directory because the unit tests were creating and moving files within it (and thus causing permissions issues in the tests). I bet a more comprehensive solution or more encompassing chmod would be helpful

TimothyBJacobs commented:
Unfortunately my first attempt at chmoding didn't seem to work. Trying it from a different place. I'm not sure I totally understand what the actual directory structure is.

I think the tests need permission on `wp-content/*`.

noahtallen commented:
the directory structure is something like. sorry if this is obvious 😅 

```
/var/www/html # WordPress root
  /wp-content
    /plugins
      /gutenberg # Via a mapped docker volume
        /build # webpack copies JS/CSS build files here
```

I think in travis, cwd is `/gutenberg`

noahtallen commented:
And I think more importantly, `chmod` is probably operating on the file structure _inside_ of Travis, whereas that file structure is actually inside the wp-env docker environment. So chmoding `./build` happens to work because it is mapped inside the volume

TimothyBJacobs commented:
Ah, I see thanks for clarifying! How would we go about running `chmod` inside the container?

noahtallen commented:
You can do “wp-env run tests-cli $command”

noahtallen commented:
I’m still not sure why it would work locally if it’s something about the environment though :)

TimothyBJacobs commented:
The only difference I can see is that locally I'm not doing the `npm run env connect` flow with a separate wordpress develop download.

TimothyBJacobs commented:
It looks like that chmod fails with `chmod: wp-content: Operation not permitted`

noahtallen commented:
> I'm not doing the npm run env connect flow with a separate wordpress develop download.

Yeah. I guess that could be related, but we are slowly switching things to wp-env, so I wouldn’t have thought that any of the Wordpress develop stuff would relate to wp-env... oh I guess one thing is that that Wordpress develop checkout in Travis would be in a subdirectory of Gutenberg, so the files would exist inside Docker. Not sure why those files would be loaded from wp-env though.

Thanks for debugging this a bit, we definitely need to improve the setup if we can

TimothyBJacobs commented:
So it looks like `wp-content` ends up owned as `root`. I don't really understand enough of how this works to understand why that is.

```
-rw-r--r--    1 xfs      xfs            405 Feb  6 06:33 index.php
934-rw-r--r--    1 xfs      xfs          19915 Feb 12 11:54 license.txt
935-rw-r--r--    1 xfs      xfs           7278 Jan 10 14:05 readme.html
936-rw-r--r--    1 xfs      xfs           6912 Feb  6 06:33 wp-activate.php
937drwxr-xr-x    9 xfs      xfs           4096 Apr 29 18:58 wp-admin
938-rw-r--r--    1 xfs      xfs            351 Feb  6 06:33 wp-blog-header.php
939-rw-r--r--    1 xfs      xfs           2275 Feb  6 06:33 wp-comments-post.php
940-rw-r--r--    1 xfs      xfs           2823 May 20 05:22 wp-config-sample.php
941-rw-r--r--    1 xfs      xfs           3189 May 20 05:22 wp-config.php
942drwxr-xr-x    5 root     root          4096 May 20 05:22 wp-content
943-rw-r--r--    1 xfs      xfs           3940 Feb  6 06:33 wp-cron.php
944drwxr-xr-x   21 xfs      xfs          12288 Apr 29 18:58 wp-includes
945-rw-r--r--    1 xfs      xfs           2496 Feb  6 06:33 wp-links-opml.php
946-rw-r--r--    1 xfs      xfs           3300 Feb  6 06:33 wp-load.php
947-rw-r--r--    1 xfs      xfs          47874 Feb 10 03:50 wp-login.php
948-rw-r--r--    1 xfs      xfs           8509 Apr 14 11:34 wp-mail.php
949-rw-r--r--    1 xfs      xfs          19396 Apr 10 03:59 wp-settings.php
950-rw-r--r--    1 xfs      xfs          31111 Feb  6 06:33 wp-signup.php
951-rw-r--r--    1 xfs      xfs           4755 Feb  6 06:33 wp-trackback.php
952-rw-r--r--    1 xfs      xfs           3133 Feb  6 06:33 xmlrpc.php
```

> we are slowly switching things to wp-env

Would part of that be #20280?

tellyworth committed:
Clean up block directory controller

This draws on https://github.com/WordPress/gutenberg/pull/17669 and https://github.com/tellyworth/wordpress-develop/pull/1. It fixes many small issues with the block directory controller, and adds unit tests.

noahtallen commented:
> Would part of that be #20280?

Yeah, that switches the e2e tests over.

> So it looks like wp-content ends up owned as root

This might be intentional to fix a permissions error we encountered on Linux, but I’ll have to look at the code again 👀

TimothyBJacobs committed:
Refactors block directory controller to make inner requests to the plugins controller

TimothyBJacobs commented:
I pushed a commit that refactors the block directory install and uninstall endpoints to use the plugins controller. Ideally we'll drop those methods and the block directory would `POST` to the `plugins` endpoint.

One thing of note is the uninstall endpoint. It is performing a deactivation and uninstall in one request. As far as I know this can't happen in the WordPress UI typically. It can also lead to all sorts of weird errors since the request started with the plugin active and running its code, but ends the request without the plugin files. This would mean if that plugin would include any files on a hook that occurred after it got uninstalled, everything could blow up. I've left this for now, but I think we should drop it and do the separate status update to inactive and then delete.

TimothyBJacobs committed:
Fix permission check for activating a plugin on install. Add more links.

TimothyBJacobs commented:
Also, I dropped the item specific permission checks for activating and deactivating since that will be handled by the plugins controller. The purpose of those permission checks now is just to generate an `Allow` header which wouldn't have the plugin slug available.

I've also fixed the single `activate_plugin` cap check. It was operating on the plugin slug, but is meant to operate on the whole plugin file. This means we can't preemptively check it before doing the install if we are going to activate it in the same request. But we check it once the install is complete.

TimothyBJacobs committed:
Check if the block plugin is already installed.

Applies 562e84e to the refactored block directory controller.

TimothyBJacobs committed:
Skip tests on fs errors.

TimothyBJacobs committed:
Fix PHPCS

TimothyBJacobs committed:
Refactors directory tests to use the controller base class.

This also enables experiments in the phpunit bootstrap.

TimothyBJacobs committed:
Add some more tests. Fix status handling.

TimothyBJacobs commented:
I've adjusted the tests to mark as skipped instead of erroring if they run into a permission error while we wait on #22515. The tests should all work if you run them locally.

I also started to add some multisite tests, but I couldn't get multisite tests to work using `test-unit-php-multisite`. AFAICT, it is still running the tests on a single-site setup and it seems like the `ms-required` and `ms-excluded` groups aren't respected. I've also marked those as skipped for now.

noahtallen commented:
> AFAICT, it is still running the tests on a single-site setup

Interesting 🤔 What needs to be done to fix this? The heuristic I used was the fact that wp-phpunit says it is running as multisite:

```
$ npm run test-unit-php-multisite

# etc...

Running as multisite...

# etc...

✔ Ran `WP_MULTISITE=1 phpunit -c /var/www/html/wp-content/plugins/gutenberg/phpunit.xml.dist` in 'phpunit'. (in 5s 96ms)
```

TimothyBJacobs commented:
Adding this to the test class `var_dump( is_multisite() ? 'Is multisite true' : 'Is multisite false' );`  says `is multisite false`.

I added some debugging code to the bootstrap file from `wp-phpunit`. It looks like the `MULTISITE` constant is getting set to `false` when loading the wp test config file.

Looking at the evaluated `wp-config.php` file it looks like that isn't defining the `MULTISITE` constant anywhere. It then loads `wp-settings.php` which checks `is_mulitiste` and if it returns false, then it does `define( 'MULTISITE', false );`.

The only difference that stands out to me is that core is using the `multisite.xml` file when running multisite tests.

Though now stepping through core's bootstrap, it looks like in core `wp-settings.php` isn't loaded by the test config file, and is instead loaded manually.

TimothyBJacobs committed:
Use wp_tests_options instead of manually using a filter.

TimothyBJacobs committed:
Fix multisite tests.

TimothyBJacobs committed:
Add a bunch more multisite test.

TimothyBJacobs commented:
I think this is ready for looking at again. We got the multisite test setup to work properly, and I've added a host of multisite specific tests.

There is still the file permissions error on travis, but the tests should pass for you locally. On travis I'm marking those as skipped if they encounter a file permissions error.

TimothyBJacobs committed:
Fix PHPCS

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+	public function update_item_permissions_check( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
+
+		if ( ! current_user_can( 'activate_plugins' ) ) {
```
What about network activating a plugin?

TimothyBJacobs replied:
`activate_plugins` is used as a barrier to all plugin management, that's how `wp-admin/plugins.php` works. The status checks for activating/deactivating are in `plugin_status_permission_check` which has a check for `network-active` plugins.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+		$data['_file'] = $file;
+
+		$response = $this->prepare_item_for_response( $data, $request );
+		$response->set_status( 201 );
```
301?

TimothyBJacobs replied:
`201` is correct for a created response.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+			'status'       => $this->get_plugin_status( $item['_file'] ),
+			'name'         => $item['Name'],
+			'plugin_uri'   => $item['PluginURI'],
+			'author'       => $item['Author'],
```
What about 
```
'TextDomain'  => 'Text Domain',
'DomainPath'  => 'Domain Path',
```

TimothyBJacobs replied:
As per the themes ticket, going to include text domain. Domain path is an internal detail, and I don't think we just want to expose the header for it.

TimothyBJacobs replied:
Added in a4bb8eb.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+	 * @return true|WP_Error
+	 */
+	protected function plugin_status_permission_check( $plugin, $new_status, $current_status ) {
+		if ( ( 'network-active' === $current_status || 'network-active' === $new_status ) && ! current_user_can( 'manage_network_plugins' ) ) {
```
Wrap in a `is_multisite()` check.

TimothyBJacobs replied:
Fixed in 0d44167.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+	 * @return true|WP_Error True if the request has read access, WP_Error object otherwise.
+	 */
+	public function get_items_permissions_check( $request ) { // phpcs:ignore VariableAnalysis.CodeAnalysis.VariableAnalysis.UnusedVariable
+		if ( ! current_user_can( 'activate_plugins' ) ) {
```
What about the `manage_network_plugins`? 
https://github.com/WordPress/wordpress-develop/blob/master/src/wp-admin/network/menu.php#L105

TimothyBJacobs replied:
That is handled by `map_meta_cap`, https://github.com/WordPress/wordpress-develop/blob/cc62b4ba1da5dc319e28f5538ea4936451bd3c7c/src/wp-includes/capabilities.php#L476.

See also `check_read_permission` which excludes network only plugins from being in the plugins list for users who can't manage network plugins if the plugin is not active.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,822 @@
+	 * @param WP_REST_Request $request Full details about the request.
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function create_item( $request ) {
```
Should we support uploading the plugin as file?

TimothyBJacobs replied:
I think that is something to consider in a future release.

TimothyBJacobs committed:
Wrap specific network permission checks in an is_multisite conditional.

See https://github.com/WordPress/gutenberg/pull/22454#discussion_r435145356

TimothyBJacobs committed:
Add text_domain to response.

See https://github.com/WordPress/gutenberg/pull/22454#discussion_r435144923

TimothyBJacobs committed:
Add more network only plugin tests.

TimothyBJacobs commented:
Thanks for the review @spacedmonkey! Pushed some commits to fix those two issues, and added more tests.

TimothyBJacobs commented:
@ryelle and @StevenDufresne do you think you could take a look at this from the existing front-end code perspective?

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+	 * @return array Query parameters for the collection.
+	 */
+	public function get_collection_params() {
+		return array(
```
get_items should be filterable by status similar to themes endpoint - 
https://github.com/WordPress/wordpress-develop/blob/5097db06aa0d923d3c8c1ae02432db5852afb2da/src/wp-includes/rest-api/endpoints/class-wp-rest-themes-controller.php#L667-L676

spacedmonkey replied:
You should also be able to search installed plugins.

spacedmonkey replied:
I really think that search, per_page and page should here as well.

TimothyBJacobs replied:
Good point, I've added `search` and `status` parameters in 3d1a7d6829.

How do you see `page`, and `per_page` working? If we try and limit the number of items processed, ie not just do an `array_slice` after the whole processing, that code path is going to get quite messy.

If we just do an `array_slice` what benefit do we get at that point? We wouldn't save any processing, just the number of items returned over the wire. And that type of pagination is very easy to implement client side, and would probably wind up as a better user experience. For reference, the plugins list table just does an `array_slice`.

I also think the chances where someone doesn't want the full list of plugins is very rare. It also seems unlikely to me that someone would have enough plugins, thousands IMO, where it would be a real issue.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+					'callback'            => array( $this, 'create_item' ),
+					'permission_callback' => array( $this, 'create_item_permissions_check' ),
+					'args'                => array(
+						'slug'   => array(
```
Should we valid the slug here? No spaces or funny characters?

TimothyBJacobs replied:
Yeah, makes sense to be safer here. Fixed in f525a35.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+					'plugin'  => array(
+						'type'              => 'string',
+						'pattern'           => self::PATTERN,
+						'validate_callback' => static function ( $file ) {
```
Let's make this methods and not static functions. It is improve testabilitiy.

TimothyBJacobs replied:
Fixed in f525a35.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+
+							return 0 === $validated;
+						},
+						'sanitize_callback' => static function ( $file ) {
```
Make this a method not a static function.

TimothyBJacobs replied:
Fixed in f525a35.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+			}
+		}
+
+		$path          = WP_PLUGIN_DIR . '/' . $file;
```
Should this be `wp_normalize_path(WP_PLUGIN_DIR)`

TimothyBJacobs replied:
No, that's how core builds that path. See `wp_get_active_and_valid_plugins` for example.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+
+		$response = $this->prepare_item_for_response( $data, $request );
+		$response->set_status( 201 );
+		$response->header( 'Location', rest_url( $this->namespace . '/' . $this->rest_base . '/' . substr( $file, 0, - 4 ) ) );
```
```suggestion
		$response->header( 'Location', rest_url( sprintf( '%s/%s/%s',  $this->namespace, $this->rest_base,  substr( $file, 0, - 4 )));
```

TimothyBJacobs replied:
Fixed in 3737daf82e.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+	 * @return WP_REST_Response|WP_Error Response object on success, or WP_Error object on failure.
+	 */
+	public function get_item( $request ) {
+		require_once ABSPATH . 'wp-admin/includes/plugin.php';
```
This line is repeated ALOT. How about a helper method?

TimothyBJacobs replied:
There are three different forms of file including. For the cases where we do repeat file inclusion, it is only of the one file form. So I don't think we get much benefit from introducing a separate method for this. See also `WP_REST_Attachments_Controller`.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+		if ( is_wp_error( $deleted ) ) {
+			return $deleted;
+		}
+
```
The resonse should look more like this -
https://github.com/WordPress/wordpress-develop/blob/5097db06aa0d923d3c8c1ae02432db5852afb2da/src/wp-includes/rest-api/endpoints/class-wp-rest-terms-controller.php#L692-L698

TimothyBJacobs replied:
Fixed in b39f30b886.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+
+		$data = $this->add_additional_fields_to_object( $data, $request );
+
+		$response = new WP_REST_Response( $data );
```
Lets break this add link to it's own method `protected function prepare_links`

TimothyBJacobs replied:
Good point, fixed in 3737daf82e.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+		$data = $this->add_additional_fields_to_object( $data, $request );
+
+		$response = new WP_REST_Response( $data );
+		$response->add_link( 'self', rest_url( $this->namespace . '/' . $this->rest_base . '/' . substr( $item['_file'], 0, - 4 ) ) );
```
```suggestion
		$response->add_link( 'self', rest_url( sprintf( '%s/%s/%s', $this->namespace, $this->rest_base, substr( $item['_file'], 0, - 4 ) ) );
```

TimothyBJacobs replied:
Fixed in 3737daf82e.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+<?php
```
What about mu-plugins?

TimothyBJacobs replied:
I don't think we should support mu-plugins or drop-ins at the moment. They can be added at a later point.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+						'status' => array(
+							'description' => __( 'The plugin activation status.', 'gutenberg' ),
+							'type'        => 'string',
+							'enum'        => is_multisite() ? array( 'inactive', 'active', 'network-active' ) : array( 'inactive', 'active' ),
```
Should there be a `network-inactive`?

TimothyBJacobs replied:
I think that just complicates the API surface. How would `network-inactive` be different from `inactive`?

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+			return $filesystem_available;
+		}
+
+		$deleted = delete_plugins( array( $request['plugin'] ) );
```
Should we also have a param here to allow you to also uninstall the plugin at the same time.

TimothyBJacobs replied:
No, that can break things. See the second half of my comment here: https://github.com/WordPress/gutenberg/pull/22454#issuecomment-631849871

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+		$result = $upgrader->install( $api->download_link );
+
+		if ( is_wp_error( $result ) ) {
+			return $result;
```
Should this add a status code?

TimothyBJacobs replied:
You're right, better to be explicit about the internal error code. Fixed in 98deefe1ab.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+
+		// This should be the same as $result above.
+		if ( is_wp_error( $skin->result ) ) {
+			return $skin->result;
```
Should this add a status code?

TimothyBJacobs replied:
Fixed in 98deefe1ab.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+			$can_change_status = $this->plugin_status_permission_check( $file, $request['status'], 'inactive' );
+
+			if ( is_wp_error( $can_change_status ) ) {
+				return $can_change_status;
```
Should this add a status code?

TimothyBJacobs replied:
Not this one, `can_change_status` is from an internal method so it generates an appropriate status code already.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+
+			$changed_status = $this->handle_plugin_status( $file, $request['status'], 'inactive' );
+
+			if ( is_wp_error( $changed_status ) ) {
```
Should this add a status code?

TimothyBJacobs replied:
Not this one, `can_change_status` is from an internal method so it generates an appropriate status code already.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+		$response = new WP_REST_Response( $data );
+		$response->add_link( 'self', rest_url( $this->namespace . '/' . $this->rest_base . '/' . substr( $item['_file'], 0, - 4 ) ) );
+
+		return $response;
```
How about a filter here 
```return apply_filters( "rest_prepare_plugin", $response, $item, $request );```

TimothyBJacobs replied:
Good idea, added in 09a21d044b.

spacedmonkey commented on lib/class-wp-rest-plugins-controller.php:
```diff
@@ -0,0 +1,829 @@
+			return true;
+		}
+
+		return new WP_Error( 'fs_unavailable', __( 'The filesystem is currently unavailable for managing plugins.', 'gutenberg' ) );
```
Status code here?

TimothyBJacobs replied:
Nice catch, fixed in 98deefe1ab.

spacedmonkey commented:
Review complete.

TimothyBJacobs committed:
Follow core pattern for DELETE responses.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437730053

TimothyBJacobs committed:
Introduce prepare_links method for generating links. Use sprintf for route building.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437730885
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437731286
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437727852

TimothyBJacobs committed:
Improve argument sanitization & validation.

Introduces a pattern to validate the .org plugin slug,
and moves specific callbacks to a dedicated method instead
of an inline closure.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437721278
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437721827
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437721999

TimothyBJacobs committed:
Explicitly add a 500 status code when returning a passed-through error.

Fix ordering issue of network only plugin activation check.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437739668
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437739714
https://github.com/WordPress/gutenberg/pull/22454#discussion_r437741613

TimothyBJacobs committed:
Add "rest_prepare_plugin" filter.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437741088

TimothyBJacobs committed:
Add "status" and "search" parameters.

https://github.com/WordPress/gutenberg/pull/22454#discussion_r437720111

TimothyBJacobs commented:
Thanks for the review @spacedmonkey! I've pushed commits or left comments for all of your feedback.

spacedmonkey reviewed (approved):
I think it is good first version. Once merged we can do some testing and tweak as required.

TimothyBJacobs commented:
> Once merged we can do some testing and tweak as required.

Definitely.

I've tested with the inserter and searching blocks and installing blocks all still works as normal in my testing. Going to go ahead and merge this.

StevenDufresne reviewed (commented):
> > Once merged we can do some testing and tweak as required.
> 
> Definitely.
> 
> I've tested with the inserter and searching blocks and installing blocks all still works as normal in my testing. Going to go ahead and merge this.

I am able to reproduce the issue that i mentioned in my review in master now. I'm curious, how did you test?

![error](https://d.pr/i/ZegyLy.png)

![error](https://d.pr/i/GEy5gD.png)

TimothyBJacobs commented on lib/class-wp-rest-block-directory-controller.php:
```diff
@@ -35,278 +35,443 @@ public function register_routes() {
-			/* translators: %s: Human-readable time difference. */
-			__( '%s ago', 'gutenberg' ),
-			human_time_diff( strtotime( $plugin['last_updated'] ), time() )
+		$query_params['term'] = array(
```
> I am able to reproduce the issue that i mentioned in my review in master now. I'm curious, how did you test?

Ack! I'm not seeing your review, did I miss it? Sorry about that, I'll work on a fix.

I tested by opening the block inserter, searching for `p3` and installing the block.

ryelle replied:
I'm also seeing the `Unable to install this block.` message, I think because the JS expects the response to be `{"success": true}`. I don't think it's an issue with the API, it's something we should fix in the Block Directory code now that this has landed.

I'm not sure about the `filterValue` issue @StevenDufresne mentions, though I do see an initial request with an empty term, it just returns regular plugins, not block-plugins.

ryelle replied:
Oh, and with the "unable …" message— the plugin installs correctly, it's just not registering in the directory.

TimothyBJacobs replied:
I think I didn't properly rebase which messed up my testing. PR incoming.

TimothyBJacobs replied:
I opened #23096 to fix it. Sorry about that!
